{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# HW _ 3 : Emperical Analysis\n",
    "# Turn in for: Rahul Deshpande (deshpr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 1. Ensure the required packages are installed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>'C:/Users/radeshpa/AppData/Local/Programs/Python/Python37/Lib/site-packages/rpy2/R/win-library/3.5'</li>\n",
       "\t<li>'C:/Program Files/R/R-3.5.2/library'</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'C:/Users/radeshpa/AppData/Local/Programs/Python/Python37/Lib/site-packages/rpy2/R/win-library/3.5'\n",
       "\\item 'C:/Program Files/R/R-3.5.2/library'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'C:/Users/radeshpa/AppData/Local/Programs/Python/Python37/Lib/site-packages/rpy2/R/win-library/3.5'\n",
       "2. 'C:/Program Files/R/R-3.5.2/library'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"C:/Users/radeshpa/AppData/Local/Programs/Python/Python37/Lib/site-packages/rpy2/R/win-library/3.5\"\n",
       "[2] \"C:/Program Files/R/R-3.5.2/library\"                                                               "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Display the directory from where we are getting the packages.\n",
    ".libPaths()\n",
    "# One or more of the commands below will fail if the library is not installed.\n",
    "library(e1071)\n",
    "library(caret)\n",
    "library(class)\n",
    "library(rpart)\n",
    "#install.packages('e1071', dependencies=TRUE)\n",
    "require('e1071')\n",
    "#install.packages('caret', dependencies = TRUE)\n",
    "require(caret)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 2. Load the Gene Expression Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>PSME1</th><th scope=col>ATF1</th><th scope=col>RHEB</th><th scope=col>FOXO3</th><th scope=col>RHOA</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ACETAZOLAMIDE</th><td>-0.015159099 </td><td>-0.031470528 </td><td>-0.004733488 </td><td> 0.02591061  </td><td> 0.0056296773</td></tr>\n",
       "\t<tr><th scope=row>IRBESARTAN</th><td>-0.026811981 </td><td> 0.012151979 </td><td>-0.025550148 </td><td>-0.02401181  </td><td>-0.0106717396</td></tr>\n",
       "\t<tr><th scope=row>IPRATROPIUM BROMIDE</th><td> 0.001017958 </td><td>-0.008650622 </td><td>-0.018128698 </td><td>-0.02079971  </td><td>-0.0002722781</td></tr>\n",
       "\t<tr><th scope=row>EFAVIRENZ</th><td>-0.004398264 </td><td> 0.055387992 </td><td> 0.004658520 </td><td> 0.01380732  </td><td>-0.0340697348</td></tr>\n",
       "\t<tr><th scope=row>THIAMINE</th><td> 0.001838965 </td><td>-0.018079188 </td><td>-0.011855532 </td><td>-0.03705033  </td><td>-0.0133954959</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllll}\n",
       "  & PSME1 & ATF1 & RHEB & FOXO3 & RHOA\\\\\n",
       "\\hline\n",
       "\tACETAZOLAMIDE & -0.015159099  & -0.031470528  & -0.004733488  &  0.02591061   &  0.0056296773\\\\\n",
       "\tIRBESARTAN & -0.026811981  &  0.012151979  & -0.025550148  & -0.02401181   & -0.0106717396\\\\\n",
       "\tIPRATROPIUM BROMIDE &  0.001017958  & -0.008650622  & -0.018128698  & -0.02079971   & -0.0002722781\\\\\n",
       "\tEFAVIRENZ & -0.004398264  &  0.055387992  &  0.004658520  &  0.01380732   & -0.0340697348\\\\\n",
       "\tTHIAMINE &  0.001838965  & -0.018079188  & -0.011855532  & -0.03705033   & -0.0133954959\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | PSME1 | ATF1 | RHEB | FOXO3 | RHOA |\n",
       "|---|---|---|---|---|---|\n",
       "| ACETAZOLAMIDE | -0.015159099  | -0.031470528  | -0.004733488  |  0.02591061   |  0.0056296773 |\n",
       "| IRBESARTAN | -0.026811981  |  0.012151979  | -0.025550148  | -0.02401181   | -0.0106717396 |\n",
       "| IPRATROPIUM BROMIDE |  0.001017958  | -0.008650622  | -0.018128698  | -0.02079971   | -0.0002722781 |\n",
       "| EFAVIRENZ | -0.004398264  |  0.055387992  |  0.004658520  |  0.01380732   | -0.0340697348 |\n",
       "| THIAMINE |  0.001838965  | -0.018079188  | -0.011855532  | -0.03705033   | -0.0133954959 |\n",
       "\n"
      ],
      "text/plain": [
       "                    PSME1        ATF1         RHEB         FOXO3      \n",
       "ACETAZOLAMIDE       -0.015159099 -0.031470528 -0.004733488  0.02591061\n",
       "IRBESARTAN          -0.026811981  0.012151979 -0.025550148 -0.02401181\n",
       "IPRATROPIUM BROMIDE  0.001017958 -0.008650622 -0.018128698 -0.02079971\n",
       "EFAVIRENZ           -0.004398264  0.055387992  0.004658520  0.01380732\n",
       "THIAMINE             0.001838965 -0.018079188 -0.011855532 -0.03705033\n",
       "                    RHOA         \n",
       "ACETAZOLAMIDE        0.0056296773\n",
       "IRBESARTAN          -0.0106717396\n",
       "IPRATROPIUM BROMIDE -0.0002722781\n",
       "EFAVIRENZ           -0.0340697348\n",
       "THIAMINE            -0.0133954959"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "took time:  0.779546"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>438</li>\n",
       "\t<li>978</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 438\n",
       "\\item 978\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 438\n",
       "2. 978\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 438 978"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# To measure time (for debugging)\n",
    "start.time <- Sys.time()\n",
    "geneExpressionFileName  <- 'gene_expression_n438x978.txt' \n",
    "myData <- read.table(geneExpressionFileName, sep=\"\\t\", header=T)\n",
    "dimnames(myData)[[1]] <- myData[,1]\n",
    "myData <- myData[,-1]\n",
    "myData[1:5, 1:5]\n",
    "end.time <- Sys.time()\n",
    "cat(\"took time: \",end.time - start.time)\n",
    "dim(myData)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We see that we have 978 genes, and 438 examples. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 3. Load the Adverse Drug Side Effect Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table>\n",
       "<thead><tr><th></th><th scope=col>Abdominal.hernias.and.other.abdominal.wall.conditions</th><th scope=col>Abortions.and.stillbirth</th><th scope=col>Acid.base.disorders</th><th scope=col>Administration.site.reactions</th><th scope=col>Adrenal.gland.disorders</th></tr></thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ACETAZOLAMIDE</th><td>0</td><td>0</td><td>1</td><td>1</td><td>0</td></tr>\n",
       "\t<tr><th scope=row>IRBESARTAN</th><td>0</td><td>0</td><td>0</td><td>0</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>IPRATROPIUM BROMIDE</th><td>0</td><td>0</td><td>0</td><td>0</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>EFAVIRENZ</th><td>0</td><td>0</td><td>0</td><td>0</td><td>1</td></tr>\n",
       "\t<tr><th scope=row>THIAMINE</th><td>0</td><td>0</td><td>0</td><td>0</td><td>0</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "\\begin{tabular}{r|lllll}\n",
       "  & Abdominal.hernias.and.other.abdominal.wall.conditions & Abortions.and.stillbirth & Acid.base.disorders & Administration.site.reactions & Adrenal.gland.disorders\\\\\n",
       "\\hline\n",
       "\tACETAZOLAMIDE & 0 & 0 & 1 & 1 & 0\\\\\n",
       "\tIRBESARTAN & 0 & 0 & 0 & 0 & 1\\\\\n",
       "\tIPRATROPIUM BROMIDE & 0 & 0 & 0 & 0 & 1\\\\\n",
       "\tEFAVIRENZ & 0 & 0 & 0 & 0 & 1\\\\\n",
       "\tTHIAMINE & 0 & 0 & 0 & 0 & 0\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "| <!--/--> | Abdominal.hernias.and.other.abdominal.wall.conditions | Abortions.and.stillbirth | Acid.base.disorders | Administration.site.reactions | Adrenal.gland.disorders |\n",
       "|---|---|---|---|---|---|\n",
       "| ACETAZOLAMIDE | 0 | 0 | 1 | 1 | 0 |\n",
       "| IRBESARTAN | 0 | 0 | 0 | 0 | 1 |\n",
       "| IPRATROPIUM BROMIDE | 0 | 0 | 0 | 0 | 1 |\n",
       "| EFAVIRENZ | 0 | 0 | 0 | 0 | 1 |\n",
       "| THIAMINE | 0 | 0 | 0 | 0 | 0 |\n",
       "\n"
      ],
      "text/plain": [
       "                    Abdominal.hernias.and.other.abdominal.wall.conditions\n",
       "ACETAZOLAMIDE       0                                                    \n",
       "IRBESARTAN          0                                                    \n",
       "IPRATROPIUM BROMIDE 0                                                    \n",
       "EFAVIRENZ           0                                                    \n",
       "THIAMINE            0                                                    \n",
       "                    Abortions.and.stillbirth Acid.base.disorders\n",
       "ACETAZOLAMIDE       0                        1                  \n",
       "IRBESARTAN          0                        0                  \n",
       "IPRATROPIUM BROMIDE 0                        0                  \n",
       "EFAVIRENZ           0                        0                  \n",
       "THIAMINE            0                        0                  \n",
       "                    Administration.site.reactions Adrenal.gland.disorders\n",
       "ACETAZOLAMIDE       1                             0                      \n",
       "IRBESARTAN          0                             1                      \n",
       "IPRATROPIUM BROMIDE 0                             1                      \n",
       "EFAVIRENZ           0                             1                      \n",
       "THIAMINE            0                             0                      "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<ol class=list-inline>\n",
       "\t<li>438</li>\n",
       "\t<li>232</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 438\n",
       "\\item 232\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 438\n",
       "2. 232\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] 438 232"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sideEffectFileName <- 'ADRs_HLGT_n438x232.txt'\n",
    "sideEffectData <- read.table(sideEffectFileName, sep=\"\\t\", header=T)\n",
    "dimnames(sideEffectData)[[1]] <- sideEffectData[,1]\n",
    "sideEffectData <- sideEffectData[,-1]\n",
    "sideEffectData[1:5, 1:5]\n",
    "dim(sideEffectData)\n",
    "names <- names(sideEffectData)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We see we have 232 side effects and 438 examples. So if we combine this with the gene data, it is like saying we have 438 examples,\n",
    "####  and 232 different  outcome variables."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 4. Some more data refining\n",
    "\n",
    "#### Make sure there are no NAs in the data, otherwise imputing might be required."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "FALSE"
      ],
      "text/latex": [
       "FALSE"
      ],
      "text/markdown": [
       "FALSE"
      ],
      "text/plain": [
       "[1] FALSE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "FALSE"
      ],
      "text/latex": [
       "FALSE"
      ],
      "text/markdown": [
       "FALSE"
      ],
      "text/plain": [
       "[1] FALSE"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "anyNA(myData)\n",
    "anyNA(sideEffectData)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We see there are no NAs in the data. Therefore, we can begin with performing our experiments."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Step 5. Experiments\n",
    "\n",
    "### We now begin performing experiments on our data. Each cell block below performs a different experiment, \n",
    "### and outputs the results to a .csv file - this helps with recording results  for the final .csv file."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Experiment 5.1, t-test with p-value < 0.03, SVM Classification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start.time <- Sys.time()\n",
    "sideEffectNames <- names(sideEffectData)\n",
    "geneNames <- names(myData)\n",
    "experimentName <- \"t_0_03_svm\"\n",
    "experimentResults <- data.frame(a = character(), b= numeric())\n",
    "experimentResults <- rbind(experimentResults, data.frame(a = \"side effect name\", b = 0))\n",
    "\n",
    "################################### EXPERIMENT OUTPUT FILE NAME ################################################################\n",
    "fileName <- paste(experimentName, '.csv')\n",
    "################################################################################################################################\n",
    "\n",
    "experimentFile <- paste(\"C:\\\\Users\\\\radeshpa\\\\Desktop\\\\BioInformatics\\\\Assignment3\\\\Experiment_output\", fileName)\n",
    "\n",
    "for(i in 1:length(sideEffectNames)){\n",
    "    sideEffectName <- sideEffectNames[i]\n",
    "    cat(\"Working on the side effect: \", sideEffectName, \"\\n\")\n",
    "    currentSideEffect <- sideEffectData[,sideEffectName]\n",
    "    presentIndices <- which(currentSideEffect == 1)\n",
    "    absentIndices <- which(currentSideEffect == 0)\n",
    "   \n",
    "    outerCount <- 3\n",
    "    accuracyOverFoldsTotal <- 0\n",
    "    \n",
    "    for(j in 1:outerCount){\n",
    "        \n",
    "        \n",
    "################################################  DATA  #######################################################################\n",
    "        \n",
    "        foldCount <- 10\n",
    "        folds <- createFolds(currentSideEffect, foldCount, returnTrain = FALSE)\n",
    "        # Get the values for every list (indices)\n",
    "\n",
    "        foldIndices <- list()\n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "            foldIndices[[length(foldIndices)+1]] <- folds[[k]]\n",
    "        }\n",
    "\n",
    "        accuracyOverFolds <- 0\n",
    "        accuracyOverFoldsSum <- 0\n",
    "       \n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "            # Data of this fold.\n",
    "            cat(\"Running fold number: \", k, \" Count number: \", j, \"\\n\")\n",
    "\n",
    "            # We need to create a vector for indexing\n",
    "            testingDataX <- myData[foldIndices[[k]],]\n",
    "\n",
    "\n",
    "            testingDataY <- sideEffectData[foldIndices[[k]],sideEffectName]\n",
    "\n",
    "\n",
    "            remainingIndices <- (Reduce(c,foldIndices[-k]))\n",
    "\n",
    "            trainingDataX <- myData[c(remainingIndices), ]\n",
    "\n",
    "            trainingDataY <- sideEffectData[c(remainingIndices), sideEffectName]\n",
    "\n",
    "##################################################### FEATURE SELECTION : t- statistic ###########################################################            \n",
    "\n",
    "            allpValues <- sapply(geneNames, function(geneName){\n",
    "                geneData <- trainingDataX[,geneName]\n",
    "                sampleOne <-  geneData[presentIndices]\n",
    "                sampleOneLength <- length(sampleOne)\n",
    "                sampleTwo <-  geneData[absentIndices]\n",
    "                sampleTwoLength <- length(sampleTwo)\n",
    "                if(sampleOneLength == 0 | sampleTwoLength == 0 | (sampleOneLength == 1 && sampleTwoLength != 1) | (sampleOneLength != 1 && sampleTwoLength == 1)){\n",
    "                    pValue  <- 0 # we want this feature.\n",
    "                }\n",
    "                else{\n",
    "                    pValue <- t.test(sampleOne, sampleTwo)$p.value\n",
    "                }\n",
    "                pValue            \n",
    "            })\n",
    "\n",
    "            ######## THRESHOLD FOR THE STATISTIC #############\n",
    "            \n",
    "            threshold <- 0.03\n",
    "            sigpIndices <- which(allpValues < threshold)\n",
    "\n",
    "            sigpValues <- allpValues[sigpIndices]\n",
    "            featuresOfInterest <- names(trainingDataX)[sigpIndices]\n",
    "\n",
    "            refinedTrainingDataX <- trainingDataX[, featuresOfInterest]\n",
    "\n",
    "            \n",
    "########################################## SVM #######################################################################################\n",
    "\n",
    "#            cat(\"Features of interest: \", length(featuresOfInterest), \"\\n\")\n",
    "            # Now filter the test set also!\n",
    "            refinedTestingDataX <- testingDataX[,featuresOfInterest]\n",
    "            svmModel <- svm(refinedTrainingDataX, trainingDataY)\n",
    "            predictionResults <- predict(svmModel, refinedTestingDataX)\n",
    "#            cat(\"Count of prediction results: \", length(predictionResults))\n",
    "\n",
    "            probabilityThreshold <- 0.5\n",
    "            predictionResults <- round(predictionResults) # default is 0.5\n",
    "            accuracy <- (sum(predictionResults==testingDataY) + sum(is.na(predictionResults) & is.na(testingDataY))) / length(predictionResults)\n",
    "            #cat(\"Accuracy for fold:  \", i, \" is \", accuracy, \"\\n\")\n",
    "            accuracyOverFoldsSum <- accuracyOverFoldsSum + accuracy\n",
    "     \n",
    "#######################################################################################################################        \n",
    "\n",
    "        }\n",
    "        \n",
    "        accuracyOverFolds <- (accuracyOverFoldsSum)/length(folds)     \n",
    "        accuracyOverFoldsTotal <- accuracyOverFoldsTotal + accuracyOverFolds\n",
    "    }\n",
    "    \n",
    "    finalAccuracy  <- accuracyOverFoldsTotal/outerCount\n",
    "    cat(\"Final Accuracy is: \", finalAccuracy, \"\\n\")\n",
    "    \n",
    "    write.table(data.frame(a = sideEffectName, b = finalAccuracy), experimentFile, append = TRUE, sep = \"\\t\", row.names = FALSE, col.names = FALSE)    \n",
    "}\n",
    "\n",
    "end.time <- Sys.time()\n",
    "cat(\"took time: \",end.time - start.time)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Experiment 5.2, Correlation of 0.05, -0.05, SVM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start.time <- Sys.time()\n",
    "sideEffectNames <- names(sideEffectData)\n",
    "geneNames <- names(myData)\n",
    "experimentName <- \"correlation_0_05_SVM_Accuracy\"\n",
    "experimentResults <- data.frame(a = character(), b= numeric())\n",
    "experimentResults <- rbind(experimentResults, data.frame(a = \"side effect name\", b = 0))\n",
    "\n",
    "################################### EXPERIMENT OUTPUT FILE NAME ################################################################\n",
    "fileName <- paste(experimentName, '.csv')\n",
    "################################################################################################################################\n",
    "\n",
    "experimentFile <- paste(\"C:\\\\Users\\\\radeshpa\\\\Desktop\\\\BioInformatics\\\\Assignment3\\\\Experiment_output\", fileName)\n",
    "\n",
    "for(i in 1:length(sideEffectNames)){\n",
    "    sideEffectName <- sideEffectNames[i]\n",
    "    cat(\"Working on the side effect: \", sideEffectName, \"\\n\")\n",
    "    currentSideEffect <- sideEffectData[,sideEffectName]\n",
    "    presentIndices <- which(currentSideEffect == 1)\n",
    "    absentIndices <- which(currentSideEffect == 0)\n",
    "\n",
    "   \n",
    "    outerCount <- 3\n",
    "    accuracyOverFoldsTotal <- 0\n",
    "    \n",
    "    for(j in 1:outerCount){\n",
    "        \n",
    "################################################  DATA  #######################################################################\n",
    "        \n",
    "        foldCount <- 10\n",
    "        folds <- createFolds(currentSideEffect, foldCount, returnTrain = FALSE)\n",
    "        # Get the values for every list (indices)\n",
    "\n",
    "        foldIndices <- list()\n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "            foldIndices[[length(foldIndices)+1]] <- folds[[k]]\n",
    "        }\n",
    "\n",
    "        accuracyOverFolds <- 0\n",
    "        accuracyOverFoldsSum <- 0\n",
    "       \n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "\n",
    "            # Data of this fold.\n",
    "#            cat(\"Running fold number: \", k, \" Count number: \", j, \"\\n\")\n",
    "\n",
    "            # We need to create a vector for indexing\n",
    "            testingDataX <- myData[foldIndices[[k]],]\n",
    "\n",
    "            testingDataY <- sideEffectData[foldIndices[[k]], sideEffectName]\n",
    "\n",
    "            remainingIndices <- (Reduce(c,foldIndices[-k]))\n",
    "\n",
    "            trainingDataX <- myData[c(remainingIndices), ]\n",
    "\n",
    "            trainingDataY <- sideEffectData[c(remainingIndices), sideEffectName]\n",
    "\n",
    "##################################################### FEATURE SELECTION : CORRELATION ###########################################################            \n",
    "\n",
    "            correlationResults <- apply(trainingDataX, 2, function(x){ cor(x, trainingDataY)})\n",
    "            boundary <- 0.05\n",
    "            geneIndices  <- c(which(correlationResults  > boundary), which(correlationResults < -boundary))\n",
    "\n",
    "            featuresOfInterest <- names(trainingDataX)[geneIndices]\n",
    "\n",
    "            refinedTrainingDataX <- trainingDataX[, featuresOfInterest]\n",
    "\n",
    "            \n",
    "########################################## SVM #######################################################################################\n",
    "\n",
    "#            cat(\"Features of interest: \", length(featuresOfInterest), \"\\n\")\n",
    "            # Now filter the test set also!\n",
    "            refinedTestingDataX <- testingDataX[,featuresOfInterest]\n",
    "            svmModel <- svm(refinedTrainingDataX, trainingDataY)\n",
    "            predictionResults <- predict(svmModel, refinedTestingDataX)\n",
    "#            cat(\"Count of prediction results: \", length(predictionResults))\n",
    "\n",
    "            probabilityThreshold <- 0.5\n",
    "            predictionResults <- round(predictionResults) # default is 0.5\n",
    "            accuracy <- (sum(predictionResults==testingDataY) + sum(is.na(predictionResults) & is.na(testingDataY))) / length(predictionResults)\n",
    "            #cat(\"Accuracy for fold:  \", i, \" is \", accuracy, \"\\n\")\n",
    "            accuracyOverFoldsSum <- accuracyOverFoldsSum + accuracy\n",
    "            \n",
    "#######################################################################################################################\n",
    "           \n",
    "        }\n",
    "        \n",
    "        accuracyOverFolds <- (accuracyOverFoldsSum)/length(folds)     \n",
    "        accuracyOverFoldsTotal <- accuracyOverFoldsTotal + accuracyOverFolds\n",
    "    }\n",
    "    \n",
    "    finalAccuracy  <- accuracyOverFoldsTotal/outerCount\n",
    "#    cat(\"Final Accuracy is: \", finalAccuracy, \"\\n\")\n",
    "    \n",
    "    write.table(data.frame(a = sideEffectName, b = finalAccuracy), experimentFile, append = TRUE, sep = \"\\t\", row.names = FALSE, col.names = FALSE)    \n",
    "}\n",
    "\n",
    "end.time <- Sys.time()\n",
    "cat(\"took time: \",end.time - start.time)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Experiment 5.3,  Correlation of 0.05, -0.05, knn = 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in eval(expr, envir, enclos): object 'sideEffectData' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in eval(expr, envir, enclos): object 'sideEffectData' not found\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "start.time <- Sys.time()\n",
    "sideEffectNames <- names(sideEffectData)\n",
    "geneNames <- names(myData)\n",
    "experimentName <- \"correlation_0_05_knn_10_Accuracy\"\n",
    "experimentResults <- data.frame(a = character(), b= numeric())\n",
    "experimentResults <- rbind(experimentResults, data.frame(a = \"side effect name\", b = 0))\n",
    "\n",
    "################################### EXPERIMENT OUTPUT FILE NAME ################################################################\n",
    "fileName <- paste(experimentName, '.csv')\n",
    "################################################################################################################################\n",
    "\n",
    "experimentFile <- paste(\"C:\\\\Users\\\\radeshpa\\\\Desktop\\\\BioInformatics\\\\Assignment3\\\\Experiment_output\", fileName)\n",
    "\n",
    "for(i in 1:length(sideEffectNames)){\n",
    "    sideEffectName <- sideEffectNames[i]\n",
    "    cat(\"Working on the side effect: \", sideEffectName, \"\\n\")\n",
    "    currentSideEffect <- sideEffectData[,sideEffectName]\n",
    "    presentIndices <- which(currentSideEffect == 1)\n",
    "    absentIndices <- which(currentSideEffect == 0)\n",
    "\n",
    "   \n",
    "    outerCount <- 3\n",
    "    accuracyOverFoldsTotal <- 0\n",
    "    \n",
    "    for(j in 1:outerCount){\n",
    "        \n",
    "################################################  DATA  #######################################################################\n",
    "        \n",
    "        foldCount <- 10\n",
    "        folds <- createFolds(currentSideEffect, foldCount, returnTrain = FALSE)\n",
    "        # Get the values for every list (indices)\n",
    "\n",
    "        foldIndices <- list()\n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "            foldIndices[[length(foldIndices)+1]] <- folds[[k]]\n",
    "        }\n",
    "\n",
    "        accuracyOverFolds <- 0\n",
    "        accuracyOverFoldsSum <- 0\n",
    "       \n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "\n",
    "            # Data of this fold.\n",
    "#            cat(\"Running fold number: \", k, \" Count number: \", j, \"\\n\")\n",
    "\n",
    "            # We need to create a vector for indexing\n",
    "            testingDataX <- myData[foldIndices[[k]],]\n",
    "\n",
    "            testingDataY <- sideEffectData[foldIndices[[k]], sideEffectName]\n",
    "\n",
    "            remainingIndices <- (Reduce(c,foldIndices[-k]))\n",
    "\n",
    "            trainingDataX <- myData[c(remainingIndices), ]\n",
    "\n",
    "            trainingDataY <- sideEffectData[c(remainingIndices), sideEffectName]\n",
    "\n",
    "##################################################### FEATURE SELECTION : CORRELATION ###########################################################            \n",
    "\n",
    "            correlationResults <- apply(trainingDataX, 2, function(x){ cor(x, trainingDataY)})\n",
    "            boundary <- 0.05\n",
    "            geneIndices  <- c(which(correlationResults  > boundary), which(correlationResults < -boundary))\n",
    "\n",
    "             featuresOfInterest <- names(trainingDataX)[geneIndices]\n",
    "\n",
    "            refinedTrainingDataX <- trainingDataX[, featuresOfInterest]\n",
    "\n",
    "            \n",
    "########################################## KNN, K = 10 #######################################################################################\n",
    "\n",
    "            # Now filter the test set also!\n",
    "            refinedTestingDataX <- testingDataX[,featuresOfInterest]\n",
    "            predictionResults <- knn(refinedTrainingDataX, refinedTestingDataX, factor(trainingDataY), k = 10)\n",
    "\n",
    "            accuracy <- (sum(predictionResults==testingDataY) + sum(is.na(predictionResults) & is.na(testingDataY))) / length(predictionResults)\n",
    "             accuracyOverFoldsSum <- accuracyOverFoldsSum + accuracy\n",
    "\n",
    "     \n",
    "#######################################################################################################################        \n",
    "           \n",
    "        }\n",
    "        \n",
    "        accuracyOverFolds <- (accuracyOverFoldsSum)/length(folds)     \n",
    "        accuracyOverFoldsTotal <- accuracyOverFoldsTotal + accuracyOverFolds\n",
    "    }\n",
    "    \n",
    "    finalAccuracy  <- accuracyOverFoldsTotal/outerCount\n",
    "#    cat(\"Final Accuracy is: \", finalAccuracy, \"\\n\")\n",
    "    \n",
    "    write.table(data.frame(a = sideEffectName, b = finalAccuracy), experimentFile, append = TRUE, sep = \"\\t\", row.names = FALSE, col.names = FALSE)    \n",
    "}\n",
    "\n",
    "end.time <- Sys.time()\n",
    "cat(\"took time: \",end.time - start.time)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Experiment 5.4, Correlation 0.10, -0.10, SVM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in eval(expr, envir, enclos): object 'sideEffectData' not found\n",
     "output_type": "error",
     "traceback": [
      "Error in eval(expr, envir, enclos): object 'sideEffectData' not found\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "start.time <- Sys.time()\n",
    "sideEffectNames <- names(sideEffectData)\n",
    "geneNames <- names(myData)\n",
    "experimentName <- \"correlation_0_10_SVM_Accuracy\"\n",
    "experimentResults <- data.frame(a = character(), b= numeric())\n",
    "experimentResults <- rbind(experimentResults, data.frame(a = \"side effect name\", b = 0))\n",
    "\n",
    "################################### EXPERIMENT OUTPUT FILE NAME ################################################################\n",
    "fileName <- paste(experimentName, '.csv')\n",
    "################################################################################################################################\n",
    "\n",
    "experimentFile <- paste(\"C:\\\\Users\\\\radeshpa\\\\Desktop\\\\BioInformatics\\\\Assignment3\\\\Experiment_output\", fileName)\n",
    "\n",
    "for(i in 1:length(sideEffectNames)){\n",
    "    sideEffectName <- sideEffectNames[i]\n",
    "    cat(\"Working on the side effect: \", sideEffectName, \"\\n\")\n",
    "    currentSideEffect <- sideEffectData[,sideEffectName]\n",
    "    presentIndices <- which(currentSideEffect == 1)\n",
    "    absentIndices <- which(currentSideEffect == 0)\n",
    "\n",
    "   \n",
    "    outerCount <- 3\n",
    "    accuracyOverFoldsTotal <- 0\n",
    "    \n",
    "    for(j in 1:outerCount){\n",
    "        \n",
    "################################################  DATA  #######################################################################\n",
    "        \n",
    "        foldCount <- 10\n",
    "        folds <- createFolds(currentSideEffect, foldCount, returnTrain = FALSE)\n",
    "        # Get the values for every list (indices)\n",
    "\n",
    "        foldIndices <- list()\n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "            foldIndices[[length(foldIndices)+1]] <- folds[[k]]\n",
    "        }\n",
    "\n",
    "        accuracyOverFolds <- 0\n",
    "        accuracyOverFoldsSum <- 0\n",
    "       \n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "\n",
    "            # Data of this fold.\n",
    "#            cat(\"Running fold number: \", k, \" Count number: \", j, \"\\n\")\n",
    "\n",
    "            # We need to create a vector for indexing\n",
    "            testingDataX <- myData[foldIndices[[k]],]\n",
    "\n",
    "            testingDataY <- sideEffectData[foldIndices[[k]], sideEffectName]\n",
    "\n",
    "            remainingIndices <- (Reduce(c,foldIndices[-k]))\n",
    "\n",
    "            trainingDataX <- myData[c(remainingIndices), ]\n",
    "\n",
    "            trainingDataY <- sideEffectData[c(remainingIndices), sideEffectName]\n",
    "\n",
    "##################################################### FEATURE SELECTION : CORRELATION ###########################################################            \n",
    "\n",
    "            correlationResults <- apply(trainingDataX, 2, function(x){ cor(x, trainingDataY)})\n",
    "            boundary <- 0.10\n",
    "            geneIndices  <- c(which(correlationResults  > boundary), which(correlationResults < -boundary))\n",
    "\n",
    "            featuresOfInterest <- names(trainingDataX)[geneIndices]\n",
    "\n",
    "            refinedTrainingDataX <- trainingDataX[, featuresOfInterest]\n",
    "\n",
    "            \n",
    "########################################## SVM #######################################################################################\n",
    "\n",
    "#            cat(\"Features of interest: \", length(featuresOfInterest), \"\\n\")\n",
    "            # Now filter the test set also!\n",
    "            refinedTestingDataX <- testingDataX[,featuresOfInterest]\n",
    "            svmModel <- svm(refinedTrainingDataX, trainingDataY)\n",
    "            predictionResults <- predict(svmModel, refinedTestingDataX)\n",
    "#            cat(\"Count of prediction results: \", length(predictionResults))\n",
    "\n",
    "            probabilityThreshold <- 0.5\n",
    "            predictionResults <- round(predictionResults) # default is 0.5\n",
    "            accuracy <- (sum(predictionResults==testingDataY) + sum(is.na(predictionResults) & is.na(testingDataY))) / length(predictionResults)\n",
    "            #cat(\"Accuracy for fold:  \", i, \" is \", accuracy, \"\\n\")\n",
    "            accuracyOverFoldsSum <- accuracyOverFoldsSum + accuracy\n",
    "            \n",
    "#######################################################################################################################\n",
    "           \n",
    "        }\n",
    "        \n",
    "        accuracyOverFolds <- (accuracyOverFoldsSum)/length(folds)     \n",
    "        accuracyOverFoldsTotal <- accuracyOverFoldsTotal + accuracyOverFolds\n",
    "    }\n",
    "    \n",
    "    finalAccuracy  <- accuracyOverFoldsTotal/outerCount\n",
    "#    cat(\"Final Accuracy is: \", finalAccuracy, \"\\n\")\n",
    "    \n",
    "    write.table(data.frame(a = sideEffectName, b = finalAccuracy), experimentFile, append = TRUE, sep = \"\\t\", row.names = FALSE, col.names = FALSE)    \n",
    "}\n",
    "\n",
    "end.time <- Sys.time()\n",
    "cat(\"took time: \",end.time - start.time)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Experiment 5.5,  t-test with p-value < 0.03, knn = 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start.time <- Sys.time()\n",
    "sideEffectNames <- names(sideEffectData)\n",
    "geneNames <- names(myData)\n",
    "experimentName <- \"t_0_03_kNN_k_10\"\n",
    "experimentResults <- data.frame(a = character(), b= numeric())\n",
    "experimentResults <- rbind(experimentResults, data.frame(a = \"side effect name\", b = 0))\n",
    "\n",
    "################################### EXPERIMENT OUTPUT FILE NAME ################################################################\n",
    "fileName <- paste(experimentName, '.csv')\n",
    "################################################################################################################################\n",
    "\n",
    "experimentFile <- paste(\"C:\\\\Users\\\\radeshpa\\\\Desktop\\\\BioInformatics\\\\Assignment3\\\\Experiment_output\", fileName)\n",
    "\n",
    "for(i in 215:length(sideEffectNames)){\n",
    "    sideEffectName <- sideEffectNames[i]\n",
    "    cat(\"Working on the side effect: \", sideEffectName, \"\\n\")\n",
    "    currentSideEffect <- sideEffectData[,sideEffectName]\n",
    "    presentIndices <- which(currentSideEffect == 1)\n",
    "    absentIndices <- which(currentSideEffect == 0)\n",
    "   \n",
    "    outerCount <- 3\n",
    "    accuracyOverFoldsTotal <- 0\n",
    "    \n",
    "    for(j in 1:outerCount){\n",
    "        \n",
    "        \n",
    "################################################  DATA  #######################################################################\n",
    "        \n",
    "        foldCount <- 10\n",
    "        folds <- createFolds(currentSideEffect, foldCount, returnTrain = FALSE)\n",
    "        # Get the values for every list (indices)\n",
    "\n",
    "        foldIndices <- list()\n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "            foldIndices[[length(foldIndices)+1]] <- folds[[k]]\n",
    "        }\n",
    "\n",
    "        accuracyOverFolds <- 0\n",
    "        accuracyOverFoldsSum <- 0\n",
    "       \n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "            # Data of this fold.\n",
    "            cat(\"Running fold number: \", k, \" Count number: \", j, \"\\n\")\n",
    "\n",
    "            # We need to create a vector for indexing\n",
    "            testingDataX <- myData[foldIndices[[k]],]\n",
    "\n",
    "\n",
    "            testingDataY <- sideEffectData[foldIndices[[k]],sideEffectName]\n",
    "\n",
    "\n",
    "            remainingIndices <- (Reduce(c,foldIndices[-k]))\n",
    "\n",
    "            trainingDataX <- myData[c(remainingIndices), ]\n",
    "\n",
    "            trainingDataY <- sideEffectData[c(remainingIndices), sideEffectName]\n",
    "\n",
    "##################################################### FEATURE SELECTION : t- statistic ###########################################################            \n",
    "\n",
    "            allpValues <- sapply(geneNames, function(geneName){\n",
    "                geneData <- trainingDataX[,geneName]\n",
    "                sampleOne <-  geneData[presentIndices]\n",
    "                sampleOneLength <- length(sampleOne)\n",
    "                sampleTwo <-  geneData[absentIndices]\n",
    "                sampleTwoLength <- length(sampleTwo)\n",
    "                if(sampleOneLength == 0 | sampleTwoLength == 0 | (sampleOneLength == 1 && sampleTwoLength != 1) | (sampleOneLength != 1 && sampleTwoLength == 1)){\n",
    "                    pValue  <- 0 # we want this feature.\n",
    "                }\n",
    "                else{\n",
    "                    pValue <- t.test(sampleOne, sampleTwo)$p.value\n",
    "                }\n",
    "                pValue            \n",
    "            })\n",
    "\n",
    "            ######## THRESHOLD FOR THE STATISTIC #############\n",
    "            \n",
    "            threshold <- 0.03\n",
    "            sigpIndices <- which(allpValues < threshold)\n",
    "\n",
    "            sigpValues <- allpValues[sigpIndices]\n",
    "            featuresOfInterest <- names(trainingDataX)[sigpIndices]\n",
    "\n",
    "            refinedTrainingDataX <- trainingDataX[, featuresOfInterest]\n",
    "\n",
    "            \n",
    "########################################## KNN, K = 10 #######################################################################################\n",
    "\n",
    "            # Now filter the test set also!\n",
    "            refinedTestingDataX <- testingDataX[,featuresOfInterest]\n",
    "            predictionResults <- knn(refinedTrainingDataX, refinedTestingDataX, factor(trainingDataY), k = 10)\n",
    "\n",
    "            accuracy <- (sum(predictionResults==testingDataY) + sum(is.na(predictionResults) & is.na(testingDataY))) / length(predictionResults)\n",
    "            accuracyOverFoldsSum <- accuracyOverFoldsSum + accuracy\n",
    "     \n",
    "#######################################################################################################################        \n",
    "\n",
    "        }\n",
    "        \n",
    "        accuracyOverFolds <- (accuracyOverFoldsSum)/length(folds)     \n",
    "        accuracyOverFoldsTotal <- accuracyOverFoldsTotal + accuracyOverFolds\n",
    "    }\n",
    "    \n",
    "    finalAccuracy  <- accuracyOverFoldsTotal/outerCount\n",
    "    cat(\"Final Accuracy is: \", finalAccuracy, \"\\n\")\n",
    "    \n",
    "    write.table(data.frame(a = sideEffectName, b = finalAccuracy), experimentFile, append = TRUE, sep = \"\\t\", row.names = FALSE, col.names = FALSE)    \n",
    "}\n",
    "\n",
    "end.time <- Sys.time()\n",
    "cat(\"took time: \",end.time - start.time)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Experiment 5.6, t-test with p-value < 0.03, SVM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start.time <- Sys.time()\n",
    "sideEffectNames <- names(sideEffectData)\n",
    "geneNames <- names(myData)\n",
    "experimentName <- \"t_0_03_svm\"\n",
    "experimentResults <- data.frame(a = character(), b= numeric())\n",
    "experimentResults <- rbind(experimentResults, data.frame(a = \"side effect name\", b = 0))\n",
    "\n",
    "################################### EXPERIMENT OUTPUT FILE NAME ################################################################\n",
    "fileName <- paste(experimentName, '.csv')\n",
    "################################################################################################################################\n",
    "\n",
    "experimentFile <- paste(\"C:\\\\Users\\\\radeshpa\\\\Desktop\\\\BioInformatics\\\\Assignment3\\\\Experiment_output\", fileName)\n",
    "\n",
    "for(i in 1:length(sideEffectNames)){\n",
    "    sideEffectName <- sideEffectNames[i]\n",
    "    cat(\"Working on the side effect: \", sideEffectName, \"\\n\")\n",
    "    currentSideEffect <- sideEffectData[,sideEffectName]\n",
    "    presentIndices <- which(currentSideEffect == 1)\n",
    "    absentIndices <- which(currentSideEffect == 0)\n",
    "   \n",
    "    outerCount <- 3\n",
    "    accuracyOverFoldsTotal <- 0\n",
    "    \n",
    "    for(j in 1:outerCount){\n",
    "        \n",
    "        \n",
    "################################################  DATA  #######################################################################\n",
    "        \n",
    "        foldCount <- 10\n",
    "        folds <- createFolds(currentSideEffect, foldCount, returnTrain = FALSE)\n",
    "        # Get the values for every list (indices)\n",
    "\n",
    "        foldIndices <- list()\n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "            foldIndices[[length(foldIndices)+1]] <- folds[[k]]\n",
    "        }\n",
    "\n",
    "        accuracyOverFolds <- 0\n",
    "        accuracyOverFoldsSum <- 0\n",
    "       \n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "            # Data of this fold.\n",
    "            cat(\"Running fold number: \", k, \" Count number: \", j, \"\\n\")\n",
    "\n",
    "            # We need to create a vector for indexing\n",
    "            testingDataX <- myData[foldIndices[[k]],]\n",
    "\n",
    "\n",
    "            testingDataY <- sideEffectData[foldIndices[[k]],sideEffectName]\n",
    "\n",
    "\n",
    "            remainingIndices <- (Reduce(c,foldIndices[-k]))\n",
    "\n",
    "            trainingDataX <- myData[c(remainingIndices), ]\n",
    "\n",
    "            trainingDataY <- sideEffectData[c(remainingIndices), sideEffectName]\n",
    "\n",
    "##################################################### FEATURE SELECTION : t- statistic ###########################################################            \n",
    "\n",
    "            allpValues <- sapply(geneNames, function(geneName){\n",
    "                geneData <- trainingDataX[,geneName]\n",
    "                sampleOne <-  geneData[presentIndices]\n",
    "                sampleOneLength <- length(sampleOne)\n",
    "                sampleTwo <-  geneData[absentIndices]\n",
    "                sampleTwoLength <- length(sampleTwo)\n",
    "                if(sampleOneLength == 0 | sampleTwoLength == 0 | (sampleOneLength == 1 && sampleTwoLength != 1) | (sampleOneLength != 1 && sampleTwoLength == 1)){\n",
    "                    pValue  <- 0 # we want this feature.\n",
    "                }\n",
    "                else{\n",
    "                    pValue <- t.test(sampleOne, sampleTwo)$p.value\n",
    "                }\n",
    "                pValue            \n",
    "            })\n",
    "\n",
    "            ######## THRESHOLD FOR THE STATISTIC #############\n",
    "            \n",
    "            threshold <- 0.03\n",
    "            sigpIndices <- which(allpValues < threshold)\n",
    "\n",
    "            sigpValues <- allpValues[sigpIndices]\n",
    "            featuresOfInterest <- names(trainingDataX)[sigpIndices]\n",
    "\n",
    "            refinedTrainingDataX <- trainingDataX[, featuresOfInterest]\n",
    "\n",
    "            \n",
    "########################################## SVM #######################################################################################\n",
    "\n",
    "#            cat(\"Features of interest: \", length(featuresOfInterest), \"\\n\")\n",
    "            # Now filter the test set also!\n",
    "            refinedTestingDataX <- testingDataX[,featuresOfInterest]\n",
    "            svmModel <- svm(refinedTrainingDataX, trainingDataY)\n",
    "            predictionResults <- predict(svmModel, refinedTestingDataX)\n",
    "#            cat(\"Count of prediction results: \", length(predictionResults))\n",
    "\n",
    "            probabilityThreshold <- 0.5\n",
    "            predictionResults <- round(predictionResults) # default is 0.5\n",
    "            accuracy <- (sum(predictionResults==testingDataY) + sum(is.na(predictionResults) & is.na(testingDataY))) / length(predictionResults)\n",
    "            #cat(\"Accuracy for fold:  \", i, \" is \", accuracy, \"\\n\")\n",
    "            accuracyOverFoldsSum <- accuracyOverFoldsSum + accuracy\n",
    "     \n",
    "#######################################################################################################################        \n",
    "\n",
    "        }\n",
    "        \n",
    "        accuracyOverFolds <- (accuracyOverFoldsSum)/length(folds)     \n",
    "        accuracyOverFoldsTotal <- accuracyOverFoldsTotal + accuracyOverFolds\n",
    "    }\n",
    "    \n",
    "    finalAccuracy  <- accuracyOverFoldsTotal/outerCount\n",
    "    cat(\"Final Accuracy is: \", finalAccuracy, \"\\n\")\n",
    "    \n",
    "    write.table(data.frame(a = sideEffectName, b = finalAccuracy), experimentFile, append = TRUE, sep = \"\\t\", row.names = FALSE, col.names = FALSE)    \n",
    "}\n",
    "\n",
    "end.time <- Sys.time()\n",
    "cat(\"took time: \",end.time - start.time)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Experiment 5.7, No Feature Selection, SVM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start.time <- Sys.time()\n",
    "sideEffectNames <- names(sideEffectData)\n",
    "geneNames <- names(myData)\n",
    "experimentName <- \"none_SVM\"\n",
    "experimentResults <- data.frame(a = character(), b= numeric())\n",
    "experimentResults <- rbind(experimentResults, data.frame(a = \"side effect name\", b = 0))\n",
    "\n",
    "################################### EXPERIMENT OUTPUT FILE NAME ################################################################\n",
    "\n",
    "fileName <- paste(experimentName, '.csv')\n",
    "\n",
    "################################################################################################################################\n",
    "\n",
    "experimentFile <- paste(\"C:\\\\Users\\\\radeshpa\\\\Desktop\\\\BioInformatics\\\\Assignment3\\\\Experiment_output\", fileName)\n",
    "\n",
    "# For every side effect (the output/outcome),  perform 10 fold cross validation. Also perform feature selection and testing at \n",
    "# every fold.\n",
    "\n",
    "cat(\"Count of side effects is: \", length(sideEffectNames), \"\\n\")\n",
    "\n",
    "for(i in 1:length(sideEffectNames)){\n",
    "    sideEffectName <- sideEffectNames[i]\n",
    "    cat(\"Working on the side effect: \", sideEffectName, \"\\n\")\n",
    "    currentSideEffect <- sideEffectData[,sideEffectName]\n",
    "    presentIndices <- which(currentSideEffect == 1)\n",
    "    absentIndices <- which(currentSideEffect == 0)\n",
    "   \n",
    "    outerCount <- 3\n",
    "    accuracyOverFoldsTotal <- 0\n",
    "    \n",
    "    \n",
    "    for(j in 1:outerCount){\n",
    "        \n",
    "        \n",
    "################################################  DATA  #######################################################################\n",
    "        \n",
    "        foldCount <- 10\n",
    "        folds <- createFolds(currentSideEffect, foldCount, returnTrain = FALSE)\n",
    "        # Get the values for every list (indices)\n",
    "\n",
    "        foldIndices <- list()\n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "            foldIndices[[length(foldIndices)+1]] <- folds[[k]]\n",
    "        }\n",
    "\n",
    "        accuracyOverFolds <- 0\n",
    "        accuracyOverFoldsSum <- 0\n",
    "       \n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "\n",
    "            # Data of this fold.\n",
    "#            cat(\"Running fold number: \", k, \" Count number: \", j, \"\\n\")\n",
    "            \n",
    "            # Create the frames for the training data and the testing data.\n",
    "            \n",
    "            # We need to create a vector for indexing\n",
    "            testingDataX <- myData[foldIndices[[k]],]\n",
    "\n",
    "\n",
    "            testingDataY <- sideEffectData[foldIndices[[k]], sideEffectName]\n",
    "\n",
    "\n",
    "            remainingIndices <- (Reduce(c,foldIndices[-k]))\n",
    "\n",
    "            trainingDataX <- myData[c(remainingIndices), ]\n",
    "\n",
    "            trainingDataY <- sideEffectData[c(remainingIndices), sideEffectName]\n",
    "\n",
    "\n",
    "            featuresOfInterest <- names(trainingDataX)\n",
    "\n",
    "            refinedTrainingDataX <- trainingDataX[, featuresOfInterest]\n",
    "\n",
    "            \n",
    "########################################## SVM #######################################################################################\n",
    "\n",
    "#            cat(\"Features of interest: \", length(featuresOfInterest), \"\\n\")\n",
    "            # Now filter the test set also!\n",
    "            refinedTestingDataX <- testingDataX[,featuresOfInterest]\n",
    "            svmModel <- svm(refinedTrainingDataX, trainingDataY)\n",
    "            predictionResults <- predict(svmModel, refinedTestingDataX)\n",
    "#            cat(\"Count of prediction results: \", length(predictionResults))\n",
    "\n",
    "            probabilityThreshold <- 0.5\n",
    "            predictionResults <- round(predictionResults) # default is 0.5\n",
    "            accuracy <- (sum(predictionResults==testingDataY) + sum(is.na(predictionResults) & is.na(testingDataY))) / length(predictionResults)\n",
    "            #cat(\"Accuracy for fold:  \", i, \" is \", accuracy, \"\\n\")\n",
    "            accuracyOverFoldsSum <- accuracyOverFoldsSum + accuracy\n",
    "            \n",
    "#######################################################################################################################\n",
    "           \n",
    "        }\n",
    "        \n",
    "        accuracyOverFolds <- (accuracyOverFoldsSum)/length(folds)     \n",
    "        accuracyOverFoldsTotal <- accuracyOverFoldsTotal + accuracyOverFolds\n",
    "    }\n",
    "    \n",
    "    finalAccuracy  <- accuracyOverFoldsTotal/outerCount\n",
    "#    cat(\"Final Accuracy is: \", finalAccuracy, \"\\n\")\n",
    "    \n",
    "    write.table(data.frame(a = sideEffectName, b = finalAccuracy), experimentFile, append = TRUE, sep = \"\\t\", row.names = FALSE, col.names = FALSE)    \n",
    "}\n",
    "\n",
    "end.time <- Sys.time()\n",
    "cat(\"took time: \",end.time - start.time)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Experiment t-test with p-value < 0.05, SVM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start.time <- Sys.time()\n",
    "sideEffectNames <- names(sideEffectData)\n",
    "geneNames <- names(myData)\n",
    "experimentName <- \"t_0_05_SVM\"\n",
    "experimentResults <- data.frame(a = character(), b= numeric())\n",
    "experimentResults <- rbind(experimentResults, data.frame(a = \"side effect name\", b = 0))\n",
    "\n",
    "################################### EXPERIMENT OUTPUT FILE NAME ################################################################\n",
    "\n",
    "fileName <- paste(experimentName, '.csv')\n",
    "\n",
    "################################################################################################################################\n",
    "\n",
    "experimentFile <- paste(\"C:\\\\Users\\\\radeshpa\\\\Desktop\\\\BioInformatics\\\\Assignment3\\\\Experiment_output\", fileName)\n",
    "\n",
    "# For every side effect (the output/outcome),  perform 10 fold cross validation. Also perform feature selection and testing at \n",
    "# every fold.\n",
    "\n",
    "cat(\"Count of side effects is: \", length(sideEffectNames), \"\\n\")\n",
    "\n",
    "for(i in 1:length(sideEffectNames)){\n",
    "    sideEffectName <- sideEffectNames[i]\n",
    "    cat(\"Working on the side effect: \", sideEffectName, \"\\n\")\n",
    "    currentSideEffect <- sideEffectData[,sideEffectName]\n",
    "    presentIndices <- which(currentSideEffect == 1)\n",
    "    absentIndices <- which(currentSideEffect == 0)\n",
    "   \n",
    "    outerCount <- 3\n",
    "    accuracyOverFoldsTotal <- 0\n",
    "    \n",
    "    \n",
    "    for(j in 1:outerCount){\n",
    "        \n",
    "        \n",
    "################################################  DATA  #######################################################################\n",
    "        \n",
    "        foldCount <- 10\n",
    "        folds <- createFolds(currentSideEffect, foldCount, returnTrain = FALSE)\n",
    "        # Get the values for every list (indices)\n",
    "\n",
    "        foldIndices <- list()\n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "            foldIndices[[length(foldIndices)+1]] <- folds[[k]]\n",
    "        }\n",
    "\n",
    "        accuracyOverFolds <- 0\n",
    "        accuracyOverFoldsSum <- 0\n",
    "       \n",
    "        \n",
    "        for(k in 1:length(folds)){\n",
    "\n",
    "            # Data of this fold.\n",
    "#            cat(\"Running fold number: \", k, \" Count number: \", j, \"\\n\")\n",
    "            \n",
    "            # Create the frames for the training data and the testing data.\n",
    "            \n",
    "            # We need to create a vector for indexing\n",
    "            testingDataX <- myData[foldIndices[[k]],]\n",
    "\n",
    "\n",
    "            testingDataY <- sideEffectData[foldIndices[[k]], sideEffectName]\n",
    "\n",
    "\n",
    "            remainingIndices <- (Reduce(c,foldIndices[-k]))\n",
    "\n",
    "            trainingDataX <- myData[c(remainingIndices), ]\n",
    "\n",
    "            trainingDataY <- sideEffectData[c(remainingIndices), sideEffectName]\n",
    "\n",
    "##################################################### FEATURE SELECTION : t- statistic ###########################################################            \n",
    "\n",
    "            allpValues <- sapply(geneNames, function(geneName){\n",
    "                geneData <- trainingDataX[,geneName]\n",
    "                sampleOne <-  geneData[presentIndices]\n",
    "                sampleOneLength <- length(sampleOne)\n",
    "                sampleTwo <-  geneData[absentIndices]\n",
    "                sampleTwoLength <- length(sampleTwo)\n",
    "\n",
    "                if(sampleOneLength == 0 | sampleTwoLength == 0 | (sampleOneLength == 1 && sampleTwoLength != 1) | (sampleOneLength != 1 && sampleTwoLength == 1)){\n",
    "                    # this is a great feature!\n",
    "                    pValue  <- 0 # we want this feature.\n",
    "                }\n",
    "                else{\n",
    "\n",
    "                    pValue <- t.test(sampleOne, sampleTwo)$p.value\n",
    "\n",
    "                } \n",
    "                pValue            \n",
    "            })\n",
    "\n",
    "\n",
    "            threshold <- 0.05\n",
    "            sigpIndices <- which(allpValues < threshold)\n",
    "\n",
    "            sigpValues <- allpValues[sigpIndices]\n",
    "            featuresOfInterest <- names(trainingDataX)[sigpIndices]\n",
    "\n",
    "            refinedTrainingDataX <- trainingDataX[, featuresOfInterest]\n",
    "\n",
    "            \n",
    "########################################## SVM #######################################################################################\n",
    "\n",
    "#            cat(\"Features of interest: \", length(featuresOfInterest), \"\\n\")\n",
    "            # Now filter the test set also!\n",
    "            refinedTestingDataX <- testingDataX[,featuresOfInterest]\n",
    "            svmModel <- svm(refinedTrainingDataX, trainingDataY)\n",
    "            predictionResults <- predict(svmModel, refinedTestingDataX)\n",
    "#            cat(\"Count of prediction results: \", length(predictionResults))\n",
    "\n",
    "            probabilityThreshold <- 0.5\n",
    "            predictionResults <- round(predictionResults) # default is 0.5\n",
    "            accuracy <- (sum(predictionResults==testingDataY) + sum(is.na(predictionResults) & is.na(testingDataY))) / length(predictionResults)\n",
    "            #cat(\"Accuracy for fold:  \", i, \" is \", accuracy, \"\\n\")\n",
    "            accuracyOverFoldsSum <- accuracyOverFoldsSum + accuracy\n",
    "            \n",
    "#######################################################################################################################\n",
    "           \n",
    "        }\n",
    "        \n",
    "        accuracyOverFolds <- (accuracyOverFoldsSum)/length(folds)     \n",
    "        accuracyOverFoldsTotal <- accuracyOverFoldsTotal + accuracyOverFolds\n",
    "    }\n",
    "    \n",
    "    finalAccuracy  <- accuracyOverFoldsTotal/outerCount\n",
    "#    cat(\"Final Accuracy is: \", finalAccuracy, \"\\n\")\n",
    "    \n",
    "    write.table(data.frame(a = sideEffectName, b = finalAccuracy), experimentFile, append = TRUE, sep = \"\\t\", row.names = FALSE, col.names = FALSE)    \n",
    "}\n",
    "\n",
    "end.time <- Sys.time()\n",
    "cat(\"took time: \",end.time - start.time)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Questions and Answers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Q1) Here are the side effects we can predict with the highest accuracy with each approach:\n",
    "1.  t = 0.05, knn, k = 10: Gastrointestinal.tract.disorders.congenital\n",
    "2. t = 0.03, knn, k = 10: Gastrointestinal.tract.disorders.congenital\n",
    "3. t = 0.05, SVM: Gastrointestinal.tract.disorders.congenital\n",
    "4. t = 0.03, SVM: Gastrointestinal.tract.disorders.congenital\n",
    "5. none, SVM: Gastrointestinal.tract.disorders.congenital\n",
    "6. correlation = 0.10, SVM: Gastrointestinal.tract.disorders.congenital\n",
    "7. correlation = 0.05, SVM: Gastrointestinal.tract.disorders.congenital\n",
    "8. correlation = 0.05, knn, k = 10 : Gastrointestinal.tract.disorders.congenital"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Q2) Here are the side effects we can predict with the lowest accuracy with each approach:\n",
    "1.  t = 0.05, knn, k = 10: Arteriosclerosis..stenosis..vascular.insufficiency.and.necrosis\n",
    "2. t = 0.03, knn, k = 10: Heart.failures\n",
    "3. t = 0.05, SVM: Heart.failures\n",
    "4. t = 0.03, SVM: Deliria..incl.confusion.\n",
    "5. none, SVM: Heart.failures\n",
    "6. correlation = 0.10, SVM: Movement.disorders..incl.parkinsonism.\n",
    "7. correlation = 0.05, SVM: Seizures..incl.subtypes.\n",
    "8. correlation = 0.05, knn, k = 10 : Movement.disorders..incl.parkinsonism."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Q3) Yes, there are such cases. Ideally we would use the F-1 score rather than the accuracy to calculate the performance of our model."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Q4) Ideally the best model would be one that offers the best performance for every side effect. Also when we are comparing multiple models,\n",
    "we can use the RMSE metric. In this case, since I already know the accuracy of all the side effects for every model, I simply chose the model\n",
    "with the highest overall average accuracy. I think this is a good start.\n",
    "\n",
    "And the answer is the model is: t = 0.05, SVM"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Q5) Yes. I noticed that when I did not perform feature selection, then it took forever for the training and testing phases to finish. This makes sense because now  the model is soo big, since there are soo many features  - a majority of which may not even be relevant. Also if I did some further data analysis using Power BI for the data  for the side effects for which the different models did terribly (or had low accuracy), I think I can see some data patterns and understand as to why the accuracies are low for these side effects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
